ligo/haskell/src/Ligo/Typescript.hs

```diff
@@ -0,0 +1,380 @@
+          [TsType parameterName $
+            TsSum $ (\x -> let n = tsmEpName x in (n <> "." <> n)) <$> epModules]
+  writeModule fp parameterModule
+
+-- Convert a 'Notes t' structure to a list of modules containing types
+-- that wraps its various entrypoints. Each entrypoint is wrapped into
+-- its own modules.
+generateEpModule :: Notes t -> [TsModule]
+generateEpModule notes = let
+  eps = lookupEntryPointsAndTypes notes
+  epNameCaps x = T.cons (toUpper $ T.head x) (T.tail x)
+  mkModule Entrypoint {..} = let
+    capsdEp = epNameCaps epName
+    in TsModule capsdEp [] $ mkTypeFor capsdEp epType
+  in mkModule <$> eps
+
+-- Given a parent directory and a TsModule, write down the module as a
+-- Typescript source file in the parent directory.
+writeModule :: MonadIO m => FilePath -> TsModule -> m FilePath
+writeModule fp (TsModule name imports decls) = do
+  let filename = (toString name) <.> "ts"
+  let declLines = build <$> decls
+  let importsLines = build <$> imports
+  writeFile (fp </> filename) (Fmt.pretty $ Fmt.unlinesF (importsLines <> declLines))
+  pure filename
+
+-- Given a typename, and a type, generated Typescript declarations for this
+-- type and all of its fields. Pairs are coverted into Tuples or Interfaces.
+-- If there is atleast one named field in the type, it is converted into
+-- interface.  Or else into a Typescript tuple. 'Or' structures are converted
+-- into Sum types after flattening their structure.  Rest of the types are
+-- converted into corresponding Typescript type after creating type aliased for
+-- their inner fields.
+mkTypeFor :: Typename -> EpType -> [TsDecl]
+mkTypeFor typename = \case
+  EpPair af1 af2 -> let
+    fields1 = flattenPairs af1
+    fields2 = flattenPairs af2
+    allFields = fields1 <> fields2
+    in case hasAtleastOneNamedField allFields of
+      False -> let
+        allFieldTypes = snd <$> allFields
+        decls = mkTypes_ <$> (zip allFieldTypes [0..])
+        fieldTypeNames = fst <$> decls
+        fieldDefinitions = snd <$> decls
+        thisType = TsType typename (TsTuple fieldTypeNames)
+        in thisType : (concat fieldDefinitions)
+      True -> let
+        decls = mkTypesForField <$> (indexEmptyFields allFields)
+        recordFields = (\(a, b, _) -> (a, b)) <$> decls
+        thisType = TsType typename (TsInterface recordFields)
+        thisDecls = (\(_, _, c) -> c) <$> decls
+        in thisType : (concat thisDecls)
+
+  EpOr af1 af2 -> let
+    fields1 = flattenOrs af1
+    fields2 = flattenOrs af2
+
+    decls = mkTypesForField <$> (indexEmptyFields $ fields1 <> fields2)
+    typeOptions = (\(_, (b, _), _) ->  b) <$> decls
+      -- Throws away information about if the field is optional. Need to check if it is relavant in 'Or' structures
+    thisType = TsType typename (TsSum typeOptions)
+    thisDecls = (\(_, _, c) -> c) <$> decls
+    in thisType : (concat thisDecls)
+
+  EpList a -> let
+    elementTypeName = typename <> "_" <> "item"
```

Comment:
Suggestion: TypeScript normally uses CamelCase for names, maybe it would be better if we had something like `typename <> "Item"` here and in other places?

Reply:
Makes sense. Have replaced this.